src/main/java/com/openshift/internal/restclient/api/capabilities/PodExec.java

```diff
@@ -0,0 +1,177 @@
+import com.openshift.internal.restclient.URLBuilder;
+import com.openshift.internal.restclient.capability.AbstractCapability;
+import com.openshift.internal.restclient.okhttp.ResponseCodeInterceptor;
+import com.openshift.internal.restclient.okhttp.WebSocketAdapter;
+import com.openshift.restclient.IApiTypeMapper;
+import com.openshift.restclient.IClient;
+import com.openshift.restclient.capability.IStoppable;
+import com.openshift.restclient.api.capabilities.IPodExec;
+import com.openshift.restclient.http.IHttpConstants;
+import com.openshift.restclient.model.IPod;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import okhttp3.ws.WebSocket;
+import okhttp3.ws.WebSocketCall;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PodExec extends AbstractCapability implements IPodExec {
+
+	private static final Logger LOG = LoggerFactory.getLogger(IPodExec.class);
+	private static final String CAPABILITY = "exec";
+
+	private static final String COMMAND = "command";
+
+	private static final String K8S_PROTOCOL_HEADER = "X-Stream-Protocol-Version";
+
+	private static final String K8S_PROTOCOL = "channel.k8s.io";
+
+	public static final int CHANNEL_STDOUT = 1;
+	public static final int CHANNEL_STDERR = 2;
+	public static final int CHANNEL_EXECERR = 3;
+
+	private final IPod pod;
+	private final DefaultClient client;
+	private final IApiTypeMapper mapper;
+
+	public PodExec(IPod pod, IClient client) {
+		super( pod, client, CAPABILITY );
+		this.pod = pod;
+		this.client = client.adapt(DefaultClient.class);
+		this.mapper = client.adapt(IApiTypeMapper.class);
+	}
+	
+	@Override
+	public String getName() {
+		return PodExec.class.getSimpleName();
+	}
+
+	@Override
+	public IStoppable start( IPodExecOutputListener listener, Options options, String... commands ) {
+
+		if ( options == null ) {
+			options = new Options();
+		}
+
+		Map<String,String> parameters = options.getMap();
+
+		OkHttpClient okClient = client.adapt(OkHttpClient.class);
+
+		URLBuilder urlBuilder = new URLBuilder(client.getBaseURL(), mapper)
+				.resource(pod)
+				.subresource(CAPABILITY)
+				.addParameters(parameters);
+
+		// The main command and all arguments are specified as 'command' parameters
+		for ( String command : commands ) {
+			urlBuilder.addParmeter( COMMAND, command );
+		}
+
+		final String endpoint = urlBuilder.websocket();
+
+		Request request = client.newRequestBuilderTo(endpoint, IHttpConstants.MEDIATYPE_ANY)
+				.method( "GET", null )
+				.addHeader(K8S_PROTOCOL_HEADER, K8S_PROTOCOL )
+				// Unless we mark this as ignored, exceptions triggered by interceptor would be lost in dispatcher thread
+				.tag( new ResponseCodeInterceptor.Ignore() {} )
+				.build();
+
+		WebSocketCall call = WebSocketCall.create(okClient, request);
```

Comment:
just a tangential observation for @jupierce - the websocket stuff is what the pod log retrieval stuff uses ...the jenkins-plugin started leveraging pod log retrieval from the restclient a month or 2 ago.